src/engine/strat_engine/backstore/metadata/bda.rs

```diff
@@ -344,31 +395,27 @@ impl StaticHeader {
                         } else if loc_1.initialization_time > loc_2.initialization_time {
                             // If the first header block is newer, overwrite second with
                             // contents of first.
-                            BDA::write(f, &buf_loc_1, MetadataLocation::Second)?;
-                            Ok(Some(loc_1))
+                            bda_write_check(f, &buf_loc_1, MetadataLocation::Second, loc_1)
                         } else {
                             // The second header block must be newer, so overwrite first
                             // with contents of second.
-                            BDA::write(f, &buf_loc_2, MetadataLocation::First)?;
-                            Ok(Some(loc_2))
+                            bda_write_check(f, &buf_loc_2, MetadataLocation::First, loc_2)
                         }
                     }
                     (None, None) => Ok(None),
                     (Some(loc_1), None) => {
                         // Copy 1 has valid Stratis BDA, copy 2 has no magic, re-write copy 2
-                        BDA::write(f, &buf_loc_1, MetadataLocation::Second)?;
-                        Ok(Some(loc_1))
+                        bda_write_check(f, &buf_loc_1, MetadataLocation::Second, loc_1)
                     }
                     (None, Some(loc_2)) => {
                         // Copy 2 has valid Stratis BDA, copy 1 has no magic, re-write copy 1
-                        BDA::write(f, &buf_loc_2, MetadataLocation::First)?;
-                        Ok(Some(loc_2))
+                        bda_write_check(f, &buf_loc_2, MetadataLocation::First, loc_2)
                     }
                 },
                 (Ok(loc_1), Err(loc_2)) => {
-                    if loc_1.is_some() {
-                        BDA::write(f, &buf_loc_1, MetadataLocation::Second)?;
-                        Ok(loc_1)
+                    // Re-write copy 2
+                    if let Some(loc_1) = loc_1 {
```

Comment:
Thanks for introducing the ```bda_write_check``` abstraction. It opens up opportunities for some further abstraction as this match arm and the one below are virtually identical. I made an attempt at an abstraction:

```
        // Action to take if there appeared to be one malformed sigblock
        // on the device.
        // If the other sigblock appears not to exist at all, return an error.
        // If the other sigblock exists, attempt a repair of the malformed
        // sigblock and return the other sigblock.
        // sh_buf are the bytes of the other sigblock
        // sh is the optional other sigblock
        // sh_error is the error indicating a malformed sigblock 
        // write_location is where to write the optional repair.
        fn repair_on_sigblock_read_error<F>(
            f: &mut F,
            sh_buf: &[u8],
            sh: Option<StaticHeader>,
            sh_error: StratisError,
            write_location: MetadataLocation,
        ) -> StratisResult<Option<SetupResult>>
        where
            F: Read + Seek + SyncAll,
        {
            match sh {
                Some(sh) => bda_write_check(f, sh_buf, write_location, sh),
                None => Err(sh_error),
            }
        }
```

which seems to work when used here. Please try that out, or something similar.